### PR TITLE
[components] Dbt project with templating

### PR DESCRIPTION
**Pull request by schrockn**
## Summary & Motivation

This PR adds templating at the field-level in components, starting with the "translator" field in the DbtProjectComponent.

e.g.

```yaml
component_type: dbt_project
component_params:
  dbt:
    project_dir: jaffle_shop
  translator:
    key: "some_prefix/{{node.name}}"
```

## How I Tested These Changes

BK

### CODE REVIEW

**Comment by schrockn** [comment]
> [!WARNING]
> <b>This pull request is not mergeable via GitHub because a downstack PR is open. Once all requirements are satisfied, merge this PR as a stack <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/26310?utm_source=stack-comment-downstack-mergeability-warning" >on Graphite</a>.</b>
> <a href="https://graphite.dev/docs/merge-pull-requests">Learn more</a>

* **#26310** <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/26310?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈 <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/26310?utm_source=stack-comment-view-in-graphite" target="_blank">(View in Graphite)</a>
* **#26309** <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/26309?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* `master`

<h2></h2>

This stack of pull requests is managed by <a href="https://graphite.dev?utm-source=stack-comment"><b>Graphite</b></a>. Learn more about <a href="https://stacking.dev/?utm_source=stack-comment">stacking</a>.

**Review by OwenKephart — COMMENTED** [review]
I like the overall shape of this pattern, but I'm not convinced that we should fragment it out based on resource type (model_translator and seed_translator):

- Not sure that this is the only axis that people might want to modulate behavior along (e.g. changing behavior based on tags seems rather natural)
- Breaks consistency between components -- you now need to know what specific translator subtypes are available

Feels cleaner if there's just a single `translator` key -- you can put if statements inside jinja after all

**Comment by OwenKephart** [comment]
and to address the "different variables available in different contexts" bit, in this specific case I would just have `node` as the top-level object, so you'd do things like:

`key: "some_prefix/{{node.name}}"`

**Comment by schrockn** [comment]
"I like the overall shape of this pattern, but I'm not convinced that we should fragment it out based on resource type (model_translator and seed_translator):"

I am also completely unconvinced that this is a good idea. Generic translator probably better.

**Comment by OwenKephart** [comment]
To add more color to my overall supportive view, I think the translator concept is well-justified in our existing integrations as a single place to consolidate information about how to turn SystemX metadata into the relevant Dagster concepts. If this is true (aka translator is valuable and important), then in order for components to be valuable without writing python code we need to expose some of this functionality to the user in the yaml world, and this is always going to be most natural if we stick with the existing translator terminology (vs. an alternative solution that more heavily templates the yaml without going through a translator)

**Comment by schrockn** [comment]
> To add more color to my overall supportive view, I think the translator concept is well-justified in our existing integrations as a single place to consolidate information about how to turn SystemX metadata into the relevant Dagster concepts. If this is true (aka translator is valuable and important), then in order for components to be valuable without writing python code we need to expose some of this functionality to the user in the yaml world, and this is always going to be most natural if we stick with the existing translator terminology (vs. an alternative solution that more heavily templates the yaml without going through a translator)

Hmm I'm a bit confused. Are you advocating for a template-based translator approach as is proposed in this PR? I can't tell.

**Comment by OwenKephart** [comment]
@schrockn yep agree w/ the template approach (seems like the only realistic way to get programmatic logic into yaml)

**Review by schrockn — COMMENTED** [review]
TODO: put `var` into context to reference environment variables

**Review by schrockn — COMMENTED** [review]
Updated here. I removed the notion of "translator" entirely from the yaml spec and just call it "nodes" which is clean, but potentially ambiguous.

**Comment by schrockn** [comment]
s/nodes/translator/g

File: python_modules/libraries/dagster-components/dagster_components/impls/dbt_project.py [hunk 1]
```diff
@@ -35,14 +42,37 @@ def cli(project_path: Optional[str], init: bool) -> "DbtGenerateParams":
         return DbtGenerateParams(project_path=project_path, init=init)
 
 
+class DbtProjectComponentTranslator(DagsterDbtTranslator):
+    def __init__(
+        self,
+        *,
+        translator: Optional[DbtNodeTranslatorParams] = None,
```
Comment (OwenKephart) [inline python_modules/libraries/dagster-components/dagster_components/impls/dbt_project.py:49]: nit: maybe `translator_params` here? just feels weird to have a Translator have a `translator` property.

**Review by OwenKephart — APPROVED** [review]

File: python_modules/libraries/dagster-components/dagster_components/impls/dbt_project.py (hunk 1 above)
Comment (schrockn) [reply python_modules/libraries/dagster-components/dagster_components/impls/dbt_project.py]: agree updated

**Comment by schrockn** [comment]
### Merge activity

* **Dec 6, 5:14 PM EST**: A user started a stack merge that includes this pull request via [Graphite](https://app.graphite.dev/github/pr/dagster-io/dagster/26310).

**Pull request merged**